README
= php-leaderboard

Leaderboards backed by Redis in PHP, http://redis.io.

Builds off ideas proposed in http://blog.agoragames.com/2011/01/01/creating-high-score-tables-leaderboards-using-redis/.

== Contributing to php-leaderboard
 
* Check out the latest master to make sure the feature hasn't been implemented or the bug hasn't been fixed yet
* Check out the issue tracker to make sure someone already hasn't requested it and/or contributed it
* Fork the project
* Start a feature/bugfix branch
* Commit and push until you are happy with your contribution
* Make sure to add tests for it. This is important so I don't break it in a future version unintentionally.
* Please try not to mess with the version or history. If you want to have your own version, or is otherwise necessary, that is fine, but please isolate to its own commit so I can cherry-pick around it.

== Copyright

Copyright (c) 2011 David Czarnecki. See LICENSE.txt for further details.



CHANGELOG.markdown
# php-leaderboard 1.0.0 (in development)

 * Initial release

test/bootstrap.php
<?php

spl_autoload_register(function($class)
{
    $file = __DIR__.'/../lib/'.strtr($class, '\\', '/').'.php';
    if (file_exists($file)) {
        require $file;
        return true;
    }
});

?>

test/LeaderboardTest.php
<?php

require "{dirname(__FILE__)}/../lib/Leaderboard.php";

class LeaderboardTestSuite extends PHPUnit_Framework_TestCase {
    public $redis;

    protected function setUp() {
        $this->redis = new Redis();
        $this->redis->connect('127.0.0.1', 6379);
        $this->redis->flushDB();
    }

    protected function tearDown() { 
        $this->redis->close();
    }
    
    function testVersion() {
        $this->assertEquals('1.0.0', Leaderboard::VERSION);
    }
    
    function testConstructLeaderboardClassWithName() {
        $leaderboard = new Leaderboard('leaderboard');
        $this->assertEquals('leaderboard', $leaderboard->getLeaderboardName());
    }
    
    function testCloseLeaderboardConnection() {
        $leaderboard = new Leaderboard('leaderboard');
        $this->assertTrue($leaderboard->close());
    }
    
    function testAddMember() {
        $leaderboard = new Leaderboard('leaderboard');
        $this->assertEquals(1, $leaderboard->addMember('david', 69));
        $this->assertEquals(1, $this->redis->zSize('leaderboard'));
    }

    function testRemoveMember() {
        $leaderboard = new Leaderboard('leaderboard');
        $this->assertEquals(1, $leaderboard->addMember('david', 69));
        $this->assertEquals(1, $leaderboard->removeMember('david'));
        $this->assertEquals(0, $this->redis->zSize('leaderboard'));
    }

    function testTotalMembers() {
        $leaderboard = new Leaderboard('leaderboard');
        $this->assertEquals(1, $leaderboard->addMember('david', 69));
        $this->assertEquals(1, $leaderboard->totalMembers());
    }
    
    function testTotalPages() {
        $leaderboard = new Leaderboard('leaderboard');
        for ($i = 1; $i <= Leaderboard::DEFAULT_PAGE_SIZE + 1; $i++) {
            $leaderboard->addMember("member_{$i}", $i);
        }
        
        $this->assertEquals(2, $leaderboard->totalPages());
    }
    
    function testSetPageSize() {
        $leaderboard = new Leaderboard('leaderboard');
        for ($i = 1; $i <= Leaderboard::DEFAULT_PAGE_SIZE + 5; $i++) {
            $leaderboard->addMember("member_{$i}", $i);
        }
        
        $this->assertEquals(2, $leaderboard->totalPages());
        $leaderboard->setPageSize(10);
        $this->assertEquals(3, $leaderboard->totalPages());
    }
    
    
    function testTotalMembersInScoreRange() {
        $leaderboard = new Leaderboard('leaderboard');
        for ($i = 1; $i <= Leaderboard::DEFAULT_PAGE_SIZE + 1; $i++) {
            $leaderboard->addMember("member_{$i}", $i);
        }

        $this->assertEquals(3, $leaderboard->totalMembersInScoreRange(2, 4));
    }
    
    function testChangeScoreFor() {
        $leaderboard = new Leaderboard('leaderboard');
        
        $leaderboard->changeScoreFor('member_1', 5);
        $this->assertEquals(5, $leaderboard->scoreFor('member_1'));
        
        $leaderboard->changeScoreFor('member_1', 5);
        $this->assertEquals(10, $leaderboard->scoreFor('member_1'));

        $leaderboard->changeScoreFor('member_1', -5);
        $this->assertEquals(5, $leaderboard->scoreFor('member_1'));
    }
    
    function testCheckMember() {
        $leaderboard = new Leaderboard('leaderboard');
        
        $leaderboard->addMember('member_1', 10);
        $this->assertTrue($leaderboard->checkMember('member_1'));
        $this->assertFalse($leaderboard->checkMember('member_2'));
    }
    
    function testRankFor() {
        $leaderboard = new Leaderboard('leaderboard');
        for ($i = 1; $i <= Leaderboard::DEFAULT_PAGE_SIZE + 1; $i++) {
            $leaderboard->addMember("member_{$i}", $i);
        }
        
        $this->assertEquals(26, $leaderboard->rankFor('member_1'));
        $this->assertEquals(25, $leaderboard->rankFor('member_1', true));
    }

    function testScoreFor() {
        $leaderboard = new Leaderboard('leaderboard');
        for ($i = 1; $i <= Leaderboard::DEFAULT_PAGE_SIZE + 1; $i++) {
            $leaderboard->addMember("member_{$i}", $i);
        }
        
        $this->assertEquals(14, $leaderboard->scoreFor('member_14'));
    }
    
    function testScoreAndRankFor() {
        $leaderboard = new Leaderboard('leaderboard');
        for ($i = 1; $i <= 5; $i++) {
            $leaderboard->addMember("member_{$i}", $i);
        }
        
        $memberData = $leaderboard->scoreAndRankFor('member_1');
        $this->assertEquals('member_1', $memberData['member']);
        $this->assertEquals(1, $memberData['score']);
        $this->assertEquals(5, $memberData['rank']);
    }
    
    function testRemoveMembersInScoreRange() {
        $leaderboard = new Leaderboard('leaderboard');
        for ($i = 1; $i <= 5; $i++) {
            $leaderboard->addMember("member_{$i}", $i);
        }

        $this->assertEquals(5, $leaderboard->totalMembers());
        
        $leaderboard->addMember('cheater_1', 100);
        $leaderboard->addMember('cheater_2', 101);
        $leaderboard->addMember('cheater_3', 102);

        $this->assertEquals(8, $leaderboard->totalMembers());
        
        $leaderboard->removeMembersInScoreRange(100, 102);

        $this->assertEquals(5, $leaderboard->totalMembers());
        
        $leaders = $leaderboard->leaders(1);
        foreach ($leaders as $key => $value) {
            $this->assertLessThan(100, $value['score']);
        }
    }
    
    function testLeaders() {
        $leaderboard = new Leaderboard('leaderboard');
        for ($i = 1; $i <= Leaderboard::DEFAULT_PAGE_SIZE + 1; $i++) {
            $leaderboard->addMember("member_{$i}", $i);
        }

        $leaders = $leaderboard->leaders(1);
        $this->assertEquals(Leaderboard::DEFAULT_PAGE_SIZE, count($leaders));
        $this->assertEquals('member_26', $leaders[0]['member']);
        $this->assertEquals(26, $leaders[0]['score']);
        $this->assertEquals(1, $leaders[0]['rank']);
        
        $leaders = $leaderboard->leaders(2);
        $this->assertEquals(1, count($leaders));
        $this->assertEquals('member_1', $leaders[0]['member']);
        $this->assertEquals(1, $leaders[0]['score']);
        $this->assertEquals(26, $leaders[0]['rank']);
    }
    
    function testAroundMe() {
        $leaderboard = new Leaderboard('leaderboard');
        for ($i = 1; $i <= Leaderboard::DEFAULT_PAGE_SIZE * 3 + 1; $i++) {
            $leaderboard->addMember("member_{$i}", $i);
        }
        
        $this->assertEquals(Leaderboard::DEFAULT_PAGE_SIZE * 3 + 1, $leaderboard->totalMembers());
        
        $leadersAroundMe = $leaderboard->aroundMe('member_30');
        $this->assertEquals(Leaderboard::DEFAULT_PAGE_SIZE / 2, count($leadersAroundMe) / 2);

        $leadersAroundMe = $leaderboard->aroundMe('member_1');
        $this->assertEquals(ceil(Leaderboard::DEFAULT_PAGE_SIZE / 2 + 1), count($leadersAroundMe));

        $leadersAroundMe = $leaderboard->aroundMe('member_76');
        $this->assertEquals(Leaderboard::DEFAULT_PAGE_SIZE / 2, count($leadersAroundMe) / 2);
    }
    
    function testRankedInList() {
        $leaderboard = new Leaderboard('leaderboard');
        for ($i = 1; $i <= Leaderboard::DEFAULT_PAGE_SIZE; $i++) {
            $leaderboard->addMember("member_{$i}", $i);
        }
        
        $this->assertEquals(Leaderboard::DEFAULT_PAGE_SIZE, $leaderboard->totalMembers());
        $members = array('member_1', 'member_5', 'member_10');
        $rankedMembers = $leaderboard->rankedInList($members);
        
        $this->assertEquals(3, count($rankedMembers));
        
        $this->assertEquals(25, $rankedMembers[0]['rank']);
        $this->assertEquals(1, $rankedMembers[0]['score']);

        $this->assertEquals(21, $rankedMembers[1]['rank']);
        $this->assertEquals(5, $rankedMembers[1]['score']);

        $this->assertEquals(16, $rankedMembers[2]['rank']);
        $this->assertEquals(10, $rankedMembers[2]['score']);
    }
}

?>

test/phpunit.bootstrap.php
<?php

// Configure PHP
if (!defined('E_DEPRECATED')) {
    define('E_DEPRECATED', 8192);
}

ini_set('error_reporting', E_ALL | E_STRICT | E_DEPRECATED);
ini_set('display_errors', true);

// Register the autoloader
spl_autoload_register(
    function($class){
        require str_replace(array('_', '\\'), '/', $class) . '.php';
    }
);



lib/Leaderboard.php
<?php

class Leaderboard {
    const VERSION = '1.0.0';
    const DEFAULT_PAGE_SIZE = 25;
    const DEFAULT_HOST = 'localhost';
    const DEFAULT_PORT = 6379;
    
    private $_leaderboard_name;
    private $_redis_connection;
    private $_page_size;
    
    public function __construct($name, $host = Leaderboard::DEFAULT_HOST, $port = Leaderboard::DEFAULT_PORT, $pageSize = Leaderboard::DEFAULT_PAGE_SIZE) {
        $this->_leaderboard_name = $name;
        $this->_redis_connection = new Redis();
        $this->_redis_connection->connect($host, $port);
        
        if ($pageSize < 1) {
            $pageSize = Leaderboard::DEFAULT_PAGE_SIZE;
        }
        
        $this->_page_size = $pageSize;
    }
    
    public function getLeaderboardName() {
        return $this->_leaderboard_name;
    }
    
    public function setPageSize($pageSize) {
        if ($pageSize < 1) {
            $pageSize = Leaderboard::DEFAULT_PAGE_SIZE;
        }
        
        $this->_page_size = $pageSize;
    }

    public function close() {
        return $this->_redis_connection->close();
    }
    
    public function addMember($member, $score) {
        return $this->addMemberTo($this->_leaderboard_name, $member, $score);
    }
    
    public function addMemberTo($leaderboardName, $member, $score) {
        return $this->_redis_connection->zAdd($leaderboardName, $score, $member);
    }
    
    public function removeMember($member) {
        return $this->removeMemberFrom($this->_leaderboard_name, $member);
    }

    public function removeMemberFrom($leaderboardName, $member) {
        return $this->_redis_connection->zRem($leaderboardName, $member);
    }

    public function totalMembers() {
        return $this->totalMembersIn($this->_leaderboard_name);
    }
    
    public function totalMembersIn($leaderboardName) {
        return $this->_redis_connection->zCard($leaderboardName);
    }
    
    public function totalPages() {
        return $this->totalPagesIn($this->_leaderboard_name);
    }

    public function totalPagesIn($leaderboardName) {
        return ceil($this->totalMembersIn($leaderboardName) / $this->_page_size);
    }
    
    public function totalMembersInScoreRange($minScore, $maxScore) {
        return $this->totalMembersInScoreRangeIn($this->_leaderboard_name, $minScore, $maxScore);
    }
    
    public function totalMembersInScoreRangeIn($leaderboardName, $minScore, $maxScore) {
        return $this->_redis_connection->zCount($leaderboardName, $minScore, $maxScore);        
    }
    
    public function changeScoreFor($member, $delta) {
        return $this->changeScoreForIn($this->_leaderboard_name, $member, $delta);
    }

    public function changeScoreForIn($leaderboardName, $member, $delta) {
        return $this->_redis_connection->zIncrBy($leaderboardName, $delta, $member);
    }
    
    public function rankFor($member, $useZeroIndexForRank = false) {
        return $this->rankForIn($this->_leaderboard_name, $member, $useZeroIndexForRank);
    }
    
    public function rankForIn($leaderboardName, $member, $useZeroIndexForRank = false) {
        $rank = $this->_redis_connection->zRevRank($leaderboardName, $member);
        if ($useZeroIndexForRank == false) {
            $rank += 1;
        }
        
        return $rank;
    }

    public function scoreFor($member) {
        return $this->scoreForIn($this->_leaderboard_name, $member);
    }
    
    public function scoreForIn($leaderboardName, $member) {
        return $this->_redis_connection->zScore($leaderboardName, $member);
    }
    
    public function checkMember($member) {
        return $this->checkMemberIn($this->_leaderboard_name, $member);      
    }

    public function checkMemberIn($leaderboardName, $member) {
        return !($this->_redis_connection->zScore($leaderboardName, $member) == NULL);        
    }

    public function scoreAndRankFor($member, $useZeroIndexForRank = false) {
        return $this->scoreAndRankForIn($this->_leaderboard_name, $member, $useZeroIndexForRank);
    }
    
    public function scoreAndRankForIn($leaderboardName, $member, $useZeroIndexForRank = false) {
        $memberData = array();
        $memberData['member'] = $member;
        $memberData['score'] = $this->scoreForIn($leaderboardName, $member);
        $memberData['rank'] = $this->rankForIn($leaderboardName, $member, $useZeroIndexForRank);
        
        return $memberData;
    }
    
    public function removeMembersInScoreRange($minScore, $maxScore) {
        return $this->removeMembersInScoreRangeIn($this->_leaderboard_name, $minScore, $maxScore);
    }
    
    public function removeMembersInScoreRangeIn($leaderboardName, $minScore, $maxScore) {
        return $this->_redis_connection->zRemRangeByScore($leaderboardName, $minScore, $maxScore);
    }
    
    public function leaders($currentPage, $withScores = true, $withRank = true, $useZeroIndexForRank = false) {
        return $this->leadersIn($this->_leaderboard_name, $currentPage, $withScores, $withRank, $useZeroIndexForRank);
    }
    
    public function leadersIn($leaderboardName, $currentPage, $withScores = true, $withRank = true, $useZeroIndexForRank = false) {
        if ($currentPage < 1) {
            $currentPage = 1;
        }
        
        if ($currentPage > $this->totalPagesIn($leaderboardName)) {
            $currentPage = $this->totalPagesIn($leaderboardName);
        }
        
        $indexForRedis = $currentPage - 1;
        
        $startingOffset = ($indexForRedis * $this->_page_size);
        if ($startingOffset < 0) {
            $startingOffset = 0;
        }
        
        $endingOffset = ($startingOffset + $this->_page_size) - 1;
        
        $leaderData = $this->_redis_connection->zRevRange($leaderboardName, $startingOffset, $endingOffset, $withScores);
        if (!is_null($leaderData)) {
            return $this->massageLeaderData($leaderboardName, $leaderData, $withScores, $withRank, $useZeroIndexForRank);
        } else {
            return NULL;
        }
    }
    
    public function aroundMe($member, $withScores = true, $withRank = true, $useZeroIndexForRank = false) {
        return $this->aroundMeIn($this->_leaderboard_name, $member, $withScores, $withRank, $useZeroIndexForRank);
    }
    
    public function aroundMeIn($leaderboardName, $member, $withScores = true, $withRank = true, $useZeroIndexForRank = false) {
        $reverseRankForMember = $this->_redis_connection->zRevRank($leaderboardName, $member);
        
        $startingOffset = $reverseRankForMember - ($this->_page_size / 2);
        if ($startingOffset < 0) {
            $startingOffset = 0;
        }
        
        $endingOffset = ($startingOffset + $this->_page_size) - 1;
        
        $leaderData = $this->_redis_connection->zRevRange($leaderboardName, $startingOffset, $endingOffset, $withScores);
        if (!is_null($leaderData)) {
            return $this->massageLeaderData($leaderboardName, $leaderData, $withScores, $withRank, $useZeroIndexForRank);
        } else {
            return NULL;
        }
    }
    
    public function rankedInList($members, $withScores = true, $useZeroIndexForRank = false) {
        return $this->rankedInListIn($this->_leaderboard_name, $members, $withScores, $useZeroIndexForRank);
    }
    
    public function rankedInListIn($leaderboardName, $members, $withScores = true, $useZeroIndexForRank = false) {
        $leaderData = array();
        
        foreach ($members as $member) {
            $memberData = array();
            $memberData['member'] = $member;
            if ($withScores) {
                $memberData['score'] = $this->scoreForIn($leaderboardName, $member);
            }
            $memberData['rank'] = $this->rankForIn($leaderboardName, $member, $useZeroIndexForRank);
            
            array_push($leaderData, $memberData);
        }
        
        return $leaderData;
    }
        
    private function massageLeaderData($leaderboardName, $leaders, $withScores, $withRank, $useZeroIndexForRank) {
        $memberAttribute = true;
        $leaderData = array();
        
        $memberData = array();
        foreach ($leaders as $key => $value) {

            $memberData['member'] = $key;
            $memberData['score'] = $value;

            if ($withRank) {
                $memberData['rank'] = $this->rankForIn($leaderboardName, $key, $useZeroIndexForRank);
            }

            array_push($leaderData, $memberData);
            $memberData = array();
        }
        
        return $leaderData;
    }   
}

?>